apps/emqx/src/emqx_limiter/src/emqx_limiter_container.erl

```diff
@@ -0,0 +1,141 @@
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%--------------------------------------------------------------------
+
+-module(emqx_limiter_container).
+
+%% @doc the container of emqx_limiter_client
+%% used to merge clients of different type of limiters to simplify operations
+%% @end
+
+%% API
+-export([ new/0, new/1, new/2
+        , add/3, add_by_name/3
+        , check/3, retry/2
+        , check_list/2, retry_list/2
+        ]).
+
+-export_type([container/0, check_result/0]).
+
+-type container() :: #{ limiter_type() => undefined | client()
+                      , retry_key() => undefined | retry_context() | future()
+                      }.
+
+-type future() :: pos_integer().
+-type limiter_type() :: emqx_limiter_schema:limiter_type().
+-type client() :: emqx_limiter_client:client().
+-type retry_context() :: emqx_limiter_client:retry_context().
+-type bucket_name() :: emqx_limiter_schema:bucket_name().
+-type millisecond() :: non_neg_integer().
+-type check_result() :: {ok, container()}
+                      | {pause, millisecond(), container()}.
+
+-define(RETRY_KEY(Type), {retry, Type}).
+-type retry_key() :: ?RETRY_KEY(limiter_type()).
+
+%%--------------------------------------------------------------------
+%%  API
+%%--------------------------------------------------------------------
+-spec new() -> container().
+new() ->
+    new([]).
+
+%% @doc generate default data according to the type of limiter
+-spec new(list(limiter_type())) -> container().
+new(Types) ->
+    new(Types, #{}).
+
+%% @doc generate a container
+%% according to the type of limiter and the bucket name configuration of the limiter
+%% @end
+-spec new(list(limiter_type()), #{limiter_type() => emqx_limiter_schema:bucket_name()}) -> container().
+new(Types, Buckets) ->
+    Init = fun(Type, Acc) ->
+                   Client = emqx_limiter_server:connect(Type, Buckets),
+                   add(Type, Client, Acc)
+           end,
+    lists:foldl(Init, #{}, Types).
+
+%% @doc add the specified type of client to the container
+-spec add_by_name(limiter_type(),
+                  bucket_name() | #{limiter_type() => bucket_name()},
+                  container()) -> container().
+add_by_name(Type, Buckets, Container) ->
+    Client = emqx_limiter_server:connect(Type, Buckets),
+    add(Type, Client, Container).
+
+-spec add(limiter_type(), client(), container()) -> container().
+add(Type, Client, Container) ->
+    Container#{ Type => Client
+              , ?RETRY_KEY(Type) => undefined
+              }.
+
+%% @doc check the specified client
+-spec check(pos_integer(), limiter_type(), container()) -> check_result().
+check(Need, Type, C) ->
```

Comment:
nit, would be nice to have `C` as `Container` for readability. (Not mixing with `Client`)